### PR TITLE
Validate client cut text length

### PR DESCRIPTION
**Pull request by ppisar**
Client-provided unsigned 32-bit cut text length is passed to various
functions that expects argument of a different type.

E.g. "RFB 003.003\n\001\006\0\0\0\xff\xff\xff\xff" string sent to the
RFB server leads to 4294967295 msg.cct.length value that in turn is
interpreted as -1 by rfbReadExact() and thus uninitialized str buffer
with potentially sensitive data is passed to subsequent functions.

This patch fixes it by checking for a maximal value that still can be
processed correctly. It also corrects accepting length value of zero
(malloc(0) is interpreted on differnet systems differently).

Whether a client can make the server allocate up to 2 GB and cause
a denial of service on memory-tight systems is kept without answer.
A possible solution would be adding an arbitrary memory limit that is
deemed safe.

CVE-2018-7225
<https://github.com/LibVNC/libvncserver/issues/218>

### CODE REVIEW

**Comment by ppisar** [comment]
I added a second patch that limits the client cut length to 1 MB as it looks reasonable. Then I rebased them to the latest upstream master HEAD.

**Comment by bk138** [comment]
Thanks! I'll have a look ASAP

File: libvncserver/rfbserver.c [hunk 1]
```diff
@@ -2575,7 +2581,23 @@ rfbProcessClientNormalMessage(rfbClientPtr cl)
...
-	str = (char *)malloc(msg.cct.length);
+	/* uint32_t input is passed to malloc()'s size_t argument,
+	 * to rfbReadExact()'s int argument, to rfbStatRecordMessageRcvd()'s int
+	 * argument increased of sz_rfbClientCutTextMsg, and to setXCutText()'s int
+	 * argument. Here we check that the value fits into all of them to
+	 * prevent from misinterpretation and thus from accessing uninitialized
+	 * memory. CVE-2018-7225 */
+	/* But first to prevent from a denial-of-service by allocating to much
+	 * memory in the server, we impose a limit of 1 MB. */
+	if (msg.cct.length > 1<<20 || msg.cct.length > SIZE_MAX || msg.cct.length > INT_MAX - sz_rfbClientCutTextMsg) {
```
Comment (bk138) [inline libvncserver/rfbserver.c]: the check will never considet the other two conditions. pls remove them.

File: libvncserver/rfbserver.c [hunk 2]
```diff
@@ -2575,7 +2581,23 @@ rfbProcessClientNormalMessage(rfbClientPtr cl)
...
+	    rfbLog("rfbClientCutText: too big cut text length requested: %" PRIu32 "\n",
+		    msg.cct.length);
+	    rfbCloseClient(cl);
+	    return;
+	}
+
+	/* Allow zero-length client cut text. */
+	str = (char *)malloc(msg.cct.length ? msg.cct.length : 1);
```
Comment (bk138) [inline libvncserver/rfbserver.c]: pls use a calloc here to init the string to empty string aka \0

File: libvncserver/rfbserver.c [hunk 3]
```diff
@@ -88,6 +88,12 @@
 #include <errno.h>
 /* strftime() */
 #include <time.h>
+/* SIZE_MAX */
```
Comment (bk138) [inline libvncserver/rfbserver.c]: pls remove those. they're not neccessary if we check for 1MB max size.

**Comment by bk138** [comment]
Also, please update the PR to contain a single commit.

File: libvncserver/rfbserver.c [hunk 4]
```diff
@@ -2575,7 +2577,23 @@ rfbProcessClientNormalMessage(rfbClientPtr cl)
...
+	/* uint32_t input is passed to malloc()'s size_t argument,
+	 * to rfbReadExact()'s int argument, to rfbStatRecordMessageRcvd()'s int
+	 * argument increased of sz_rfbClientCutTextMsg, and to setXCutText()'s int
+	 * argument. Here we impose a limit of 1 MB so that the value fits
+	 * into all of the types to prevent from misinterpretation and thus
+	 * from accessing uninitialized memory (CVE-2018-7225) and also to
+	 * prevent from a denial-of-service by allocating to much memory in
+	 * the server. */
+	if (msg.cct.length > 1<<20) {
+	    rfbLog("rfbClientCutText: too big cut text length requested: %" PRIu32 "\n",
```
Comment (bk138) [inline libvncserver/rfbserver.c]: Why not simply use %u here?

File: libvncserver/rfbserver.c [hunk 5]
```diff
@@ -88,6 +88,8 @@
 #include <errno.h>
 /* strftime() */
 #include <time.h>
+/* PRIu32 */
+#include <inttypes.h>
```
Comment (bk138) [inline libvncserver/rfbserver.c]: When using %u, we can avoid an additional include.

**Comment by ppisar** [comment]
On Thu, Mar 08, 2018 at 11:06:32AM -0800, Christian Beier wrote:
> bk138 requested changes on this pull request.
> 
> Almost there! It also seems to me that the 2nd paragraph in the commit log is obsolete?
>
It's required. This fix is about CVE-2018-7225.

> > @@ -2575,7 +2577,23 @@ rfbProcessClientNormalMessage(rfbClientPtr cl)
>  
>  	msg.cct.length = Swap32IfLE(msg.cct.length);
>  
> -	str = (char *)malloc(msg.cct.length);
> +	/* uint32_t input is passed to malloc()'s size_t argument,
> +	 * to rfbReadExact()'s int argument, to rfbStatRecordMessageRcvd()'s int
> +	 * argument increased of sz_rfbClientCutTextMsg, and to setXCutText()'s int
> +	 * argument. Here we impose a limit of 1 MB so that the value fits
> +	 * into all of the types to prevent from misinterpretation and thus
> +	 * from accessing uninitialized memory (CVE-2018-7225) and also to
> +	 * prevent from a denial-of-service by allocating to much memory in
> +	 * the server. */
> +	if (msg.cct.length > 1<<20) {
> +	    rfbLog("rfbClientCutText: too big cut text length requested: %" PRIu32 "\n",
> 
> Why not simply use %u here?
> 
Because msg.cct.length is not of type unsigned int.

**Pull request merged**